### DISCUSSION TITLE
Having an Angular CRUD example would ne bice

### DISCUSSION THREAD

**Discussion by colorcube** [state-adapt/state-adapt#33] [General]
Having a best practice example with typical CRUD actions would be nice.

Should include

- (faked) http entity api service
- loading indicator
- error handling
- using entity adapter  (mentioned here: https://dev.to/this-is-learning/exciting-possibilities-with-state-adapters-3cia)

I found these

- https://github.com/joshuamorony/quicklist-state-adapt
- https://github.com/mfp22/state-adapt-todo-mvc

Are there any more recent examples? They are a bit outdated and/or not quite what I was looking for.

**Comment by mfp22** [comment]
You're right. I need to refresh some examples and add them to the docs. These 2 look like they will have what you want. I actually forgot to push a commit that updates the first one with the new HTTP utilities. I'll do that in the next few hours. 
- https://github.com/mfp22/2023-fidelity-workshop/tree/state-adapt
  - createEntityAdapter is used in books.adapter.ts and should be up-to-date 
- https://github.com/mfp22/stefanoslig-angular-14-ngrx-nx-realworld-example-app/tree/state-adapt

I'll add more comments here as I update these.

**Reply by colorcube** [reply]
Many thanks!
books.adapter.ts looks like what I was looking for.

An even less complex app with short (1,2 lines) comments on every important step would be nice (adapt, createEntityAdapter, filter, actions, request...).

I read your articles and saw some videos and I can follow your arguments. (Very interesting btw)
But for example Ngrx, even if you understand the principles, the code is hard to understand (nested stuff and ... operators all over the place). 
State code with StateAdapt doesn't look that 'fuzzy' but still it is a bit hard to grasp what's going on. An annotated example would be nice.

Maybe the todo example could be updated and annotated?

**Reply by mfp22** [reply]
I've updated the first one https://github.com/mfp22/2023-fidelity-workshop/tree/state-adapt

What kind of annotations? I'm not a fan of redundant or overly verbose comments. For explanations about reactivity and clean code in general, maybe I could add links to parts of articles I've written. For documentation, I could add comments with links to the docs for each StateAdapt function I'm using. You can pull the code and hover over functions in VSCode and the docs are there, but that doesn't help when looking at it in GitHub.

**Reply by colorcube** [reply]
Thanks for the update.

Regarding annotations. Something like that

```typescript
    // the http data service for crud operations on the backend
    const booksService = inject(BooksService);

    // creating request observables success$ and error$ from booksService.all() observable
    const booksRequest = getRequestSources('books', booksService.all());


    // create store with
    // - name 'books' visible in Redux DevTools
    // - initialState - the name says it
    // - adapter with state changing functions like set, reset, ...
    return adapt(['books', initialState, adapter], {
      // here are the sources for state changes
      // stateChangingFunction: Observable which emits data to that function
      receiveBooks: booksRequest.success$,
      receiveError: booksRequest.error$,
      // when bookCreated$ emit value (BookModel instance) addBook gets the value and changes state
      addBook: bookCreated$,
      updateBook: bookUpdated$ as Source<[string, BookModel]>,
      removeBooksOne: bookDeleted$,
    });
```

forgive me if I wrote nonsense :-)
I still try to understand    

Having one example that guides me a bit through the code helps me. But that is of course a matter of taste

**Reply by mfp22** [reply]
Yeah, it makes sense, but I'm really picky about comments. I hate redundant comments. Or redundant anything.

```typescript
    // the http data service for crud operations on the backend
    const booksService = inject(BooksService);
```

Is the code not clear enough? I think that line is very common in Angular apps. Services are usually used to manage API access.

```typescript
    // creating request observables success$ and error$ from booksService.all() observable
    const booksRequest = getRequestSources('books', booksService.all());
```

` from booksService.all() observable` is obvious, because we're passing in  `booksService.all()`. Which, by the way, should probably be renamed to `fetchAll()`, but I didn't worry about that. I could definitely clean up this repo... That would be helpful. If you don't know what a thing does by its name, maybe it's just a bad name. So I guess I can draw from your comments to make the code itself easier to understand. But I don't want to write comments for the sake of writing comments. _The code_ is supposed to be the description. 

`// creating request observables success$ and error$` This is something that would also be unnecessary if I had a better name than `booksRequest`. Maybe `booksRequestSources`? It's technically not itself a request... So I should rename that. 

But now do `success$` and `error$` really help to specify? What could `requestSources` be? If you understand what a `Source` is, it's very easy to imagine what `requestSources` are. But developers looking at this example may not remember what what a `Source` is. So this could be where a link to the docs would be helpful. Maybe a docs link for every StateAdapt utility would be helpful.

So altogether we would have 

```diff
    const booksService = inject(BooksService);

-   const booksRequest = getRequestSources('books', booksService.all());
+   // https://state-adapt.github.io/docs/rxjs#getrequestsources
+   const booksRequestSources = getRequestSources(
+     'books',
+     booksService.fetchAll()
+   );

    const bookCreated$ = BooksPageActions.saveBook$.pipe(
      filter(({ payload }) => !('id' in payload)),
      concatMap(({ payload }) => booksService.create(payload)),
+     // https://state-adapt.github.io/docs/rxjs#tosource
      toSource('bookCreated$')
    );

    const bookUpdated$ = BooksPageActions.saveBook$.pipe(
      filter(BooksPageActions.isBookModelAction),
      concatMap(({ payload }) =>
        booksService
          .update(payload.id, payload)
+         // https://state-adapt.github.io/docs/core#getaction
          .pipe(map((book) => getAction('bookUpdated$', [book.id, book])))
      )
    );

    const bookDeleted$ = BooksPageActions.deleteBook$.pipe(
      concatMap(({ payload }) =>
        booksService
          .delete(payload)
+         // https://state-adapt.github.io/docs/core#getaction
          .pipe(map(() => getAction('bookDeleted$', payload)))
      )
    );

+   // https://state-adapt.github.io/angular/docs/angular#overload-4
    return adapt(['books', initialState, adapter], {
-     receiveBooks: booksRequest.success$,
-     receiveError: booksRequest.error$,
+     receiveBooks: booksRequestSources.success$,
+     receiveError: booksRequestSources.error$,
      addBook: bookCreated$,
-     updateBook: bookUpdated$ as Source<[string, BookModel]>,
+     updateBook: bookUpdated$ as Source<[string, BookModel]>, // https://state-adapt.github.io/docs/rxjs#source
      removeBooksOne: bookDeleted$,
    });
```

Again, in VSCode or StackBlitz, this kind of thing would be unnecessary, because you only need to mouse-over things and the docs are embedded as JSDoc comments.

**Reply by mfp22** [reply]
Okay the repo has been updated now https://github.com/mfp22/2023-fidelity-workshop/tree/state-adapt

**Reply by mfp22** [reply]
It will take much longer to update the Real World App

**Reply by colorcube** [reply]
Having links in the docs might be even better. Thanks!

From my point of view, stores are very hard to understand. Not what they do. It's easy to understand what a coffee machine do. But if the machine is disassembled and you look at the pieces, you have no idea how it works. If each piece gets described, it becomes clearer. But you can only understand it if you have a description how all the pieces fit together.

Sorry to bother you, but that's how my brain works :-)

Nevertheless, in general StateAdapt makes totally sense to me and I would like to mention that your articles are excellent !

**Reply by mfp22** [reply]
Yeah, stores have a lot of unique pieces, that's for sure. Selectors alone are pretty complicated if you try to understand how they work. But I've tested them a lot, so they're solid.

I'm glad you like my articles. 

What do you think I should do to convince more people to use StateAdapt? More examples? Tutorials?

**Reply by colorcube** [reply]
Hard to say. In a few days I have more time to try things out.

What I think
reactive programming is a journey - unless it makes immediately click

If you're not familiar with reactive programming it's hard to make the mindset shift.
(click)="doIt()" is easy. But when you start with rxjs it's weird. And more importantly, it's work. You have to work to understand it. But it's doable.

Then comes ngrx and you're lost again. Nested functions and spread operators all over the place. I never managed to really get it. I think many people think ngrx for typical crud stuff like table view, edit, etc. is overkill. And I tend to agree, because it is hard to "read".  A very disappointing topic.

I can imagine with StateAdapt this gets better because, it is so "simple" in usage that you can easily (want to) use multiple stores in your app. You can use it for small things but also for big complex stuff. - This is my extrapolation from what I know until now. 

Showing how "easy" StateAdapt is in normal/simple scenarios (crud) might be a way to promote it.

But still it is work to get into it and people are lazy ;-)

I hope this makes some sense

**Comment by colorcube** [comment]
Could you please take a look into:
https://github.com/mfp22/2023-fidelity-workshop/pull/1/commits/df25a53153066dc31401b196d7ac498934fc353a

and a related question (based on the merge request):
If one want derived lists based on a filter, would it look like this:

books$ = inject(BooksSorted);
booksAbc$ = inject(BooksFilteredCategoryAbc);
booksXyz$ = inject(BooksFilteredCategoryXyz);

**Reply by mfp22** [reply]
The way I did stores in that project might be a little confusing. I was wanting to use a factory function instead of a class service, just so I could use `const store = ` when defining store instead of having to do class properties. I also thought it was weird defining services for stores names `SomeStoreService` and then having to access the store as a property on the class, like `someStore = inject(SomeStoreService).store`. So I was just playing around with alternative syntax I'm this project. I've decided that it's okay to use the factory functions like I did in that project if you want to make the store shared between multiple components from the beginning, but if you're creating a store in a component class and later decide to move it to an injectable, you might as well keep it as a class. But this makes inconsistent syntax based on where the store was initially created, which might bother people. So overall I would probably recommend using regular services and having a `store` property.

I would create a single store for all derived states that need to be in stores (potentially shared between multiple components). So 
```
joinedStore = inject(JoinedStore);
books$ = this.joinedStore.books$;
booksAbc$ = this.joinedStore.booksAbc$;
// etc
```

Also I wouldn't give each store's state a suffix of `Store` when you're joining stores. All of them are stores, so that's redundant. However, I see that it would create something like `booksBooksAll`, which is an awkward name.

In general I feel a bit weird about these stores being separate. The entity adapter creator actually supports sorting. So I would probably put everything in a single entity store. But for a quick addition to the project, it's fine not to refactor that. If you look at the adapter docs here https://state-adapt.github.io/adapters/core#createEntityAdapter you'll notice I specify `sorters: ['name'] ` when creating the adapter, and the selectors made available in that adapter now include some selectors ending in `byName`. Basically, you can choose whatever selectors you want to determine sorting. They just have to be in the initial adapter you give to `createEntityAdapter`.

Documentation is a challenge.

**Reply by mfp22** [reply]
Oh and another option for derived state is signals. When you call `toSignal`, it subscribes to the observable. So I would only use it in components, where you're actually using the data. Then a `computed` for the derived state. But I would only do that for derived state that will always be coupled to the component template. Because all shared signals in Angular are going to live forever. It's better to use RxJS for shared state and signals for state that can always be created and destroyed with specific components.

**Reply by colorcube** [reply]
Currently I am porting an angular app to StateAdapt from NgRx. I start with very simple stores, and now I understand the overall concept and usage better.

Best practice examples are the key, at least for me. I looked multiple times in the documentation, but most of the time I looked into the examples. From easiest to book store. Now I have a bit understanding how the StateAdapt API works to create a store, but not from the documentation. That's too abstract. I understand the docs, but not how things work together. 

I don't say the documentation is bad. It's more like StateAdapt is a toolbox to create a store and examples are needed to show how to use it the right way.

I can already say that I like StateAdapt and it's the first time I understand the store stuff I'm doing ;-)
It's way less code than NgRx. I think it's also much more flexible.

Naming is indeed a bit challenging, especially when using joinAdapters() you could get funny results :-)
But in that case the naming might have been not very good in the beginning.

**Reply by mfp22** [reply]
Yeah, usually when I struggle coming up with nice names it's a sign that something weird is going on.

I've been making a ton of YouTube videos and I have a lot of repositories with examples, but they're probably outdated, and they're not very accessible either. But I have a lot of fun using StateAdapt, so if you have any suggestions on how I can generate a ton of examples or how to link to them from the docs, that would be helpful.

I assume you're using the version of StateAdapt that integrates with NgRx? I guess that would be @state-adapt/ngrx

**Reply by colorcube** [reply]
No I don't want to integrate with ngrx. I want to get rid of it ;-)

**Reply by mfp22** [reply]
Okay, so probably not a huge app :) Otherwise you could convert one reducer at a time and the remove @state-adapt/ngrx at the end.

**Reply by colorcube** [reply]
I'm not sure if a ton of examples are needed, but 2 or 3 clean slightly more complex examples. The book store is in principle right, regarding the size and complexity. It has

- typical auth/login 
- typical CRUD with list/detail display and sort

The sorting should be fixed. Is BooksListStore needed? ...

I would like to see an example with SomeStoreService as you mentioned.

**Comment by colorcube** [comment]
> Okay, so probably not a huge app :) Otherwise you could convert one reducer at a time and the remove @state-adapt/ngrx at the end.

They both work independent of each other. 
Fun fact: replacing a ngrx store with stateadapt results in better understanding of ngrx :-)

**Reply by mfp22** [reply]
Interesting! Is it because it has all the same pieces, but with terser syntax?

**Reply by mfp22** [reply]
Oh and yeah it makes sense that they would both work. They'll just be in different Redux Devtools stores.

**Reply by colorcube** [reply]
There's no intersection. Everything regarding store A, B is changed to StateAdapt. Store C, D, E, F is still NgRx. In Redux DevTools I have two times 'My App' (instances) which confused me because I didn't see the StateAdapt stores. But you can switch between the two. Everything seems to work fine.

**Reply by mfp22** [reply]
Good. If you do want to see them in the same devtools, you can use that package I mentioned. It should be very easy to configure. I guess the drawback is that you would have to use `adaptNgrx` everywhere instead of just `adapt`. https://state-adapt.github.io/angular/get-started#angular-and-ngrx
